carl/envs/carl_env_new.py

```diff
@@ -0,0 +1,253 @@
+from carl.context.context_space import (
+    ContextFeature,
+    ContextSpace,
+    UniformFloatContextFeature,
+    ContextSampler,
+    NormalFloatContextFeature,
+)
+
+
+class CARLEnv(Wrapper, abc.ABC):
+    def __init__(
+        self,
+        env: Env,
+        contexts: Contexts | None = None,
+        state_context_features: list[str]
+        | None = None,  # list the context features which should be added to the state # TODO rename to obs_context_features?
+        obs_context_as_dict: bool = True,  # TODO discuss default
+        context_selector: AbstractSelector | type[AbstractSelector] | None = None,
+        context_selector_kwargs: dict = None,
+        **kwargs,
+    ):
+        super().__init__(env)
+
+        self.state_observation_space: gymnasium.spaces.Space = env.observation_space
+        self.obs_context_as_dict = obs_context_as_dict
+
+        self.context = None
+        if contexts is None:
+            contexts = {0: self.get_default_context(self)}
+        self.contexts = contexts
+        self.state_context_features = state_context_features
+
+        # Context Selector
+        self.context_selector: type[AbstractSelector]
+        if context_selector is None:
+            self.context_selector = RoundRobinSelector(contexts=contexts)  # type: ignore [assignment]
+        elif isinstance(context_selector, AbstractSelector):
+            self.context_selector = context_selector  # type: ignore [assignment]
+        elif inspect.isclass(context_selector) and issubclass(
+            context_selector, AbstractSelector
+        ):
+            if context_selector_kwargs is None:
+                context_selector_kwargs = {}
+            _context_selector_kwargs = {"contexts": contexts}
+            context_selector_kwargs.update(_context_selector_kwargs)
+            self.context_selector = context_selector(**context_selector_kwargs)  # type: ignore [assignment]
+        else:
+            raise ValueError(
+                f"Context selector must be None or an AbstractSelector class or instance. "
+                f"Got type {type(context_selector)}."
+            )
+        self._progress_instance()
+        self._update_context()
+        self.observation_space = self.get_observation_space(
+            obs_context_feature_names=self.state_context_features
+        )
+
+    def get_observation_space(
+        self, obs_context_feature_names: list[str] | None = None
+    ) -> gymnasium.spaces.Dict:
+        context_space = self.get_context_space()
+        obs_space_context = context_space.to_gymnasium_space(
+            context_feature_names=obs_context_feature_names,
+            as_dict=self.obs_context_as_dict,
+        )
+        obs_space = spaces.Dict(
+            {
+                "state": self.state_observation_space,
+                "context": obs_space_context,
+            }
+        )
+        return obs_space
+
+    @staticmethod
+    def get_context_features() -> list[ContextFeature]:
```

Comment:
Why list? Dict would be much nicer imo

Reply:
Same as above, let's discuss later and agree on sth